app/javascript/components/routers-form/index.jsx

```diff
@@ -0,0 +1,163 @@
+  }, setState] = useState({
+    isLoading: !!routerId, fields: [], subnetSelect: [], subnets: [],
+  });
+  const submitLabel = !!routerId ? __('Save') : __('Add');
+
+  let network = '';
+
+  const callAPI = (value) => {
+    API.get(`/api/cloud_subnets?expand=resources&attributes=name,ems_ref&filter[]=cloud_network_id=${value}`).then((data) => {
+      let subnets = data.resources;
+      subnets = subnets.map(({ id, name }) => ({ label: name, value: id }));
+      setState((state) => ({
+        ...state,
+        subnets,
+      }));
+    });
+  };
+
+  const loadSchema = (appendState = {}) => ({ data: { form_schema: { fields } } }) => {
+    Object.assign(fields[2].fields[2], {
+      resolveProps: (props, { meta, input }) => {
+        if (input.value && meta.active) {
+          if (input.value !== network) {
+            callAPI(input.value);
+            network = input.value;
+          }
+        }
+      },
+    });
+    setState((state) => ({
+      ...state,
+      ...appendState,
+      fields,
+    }));
+  };
+
+  const emptySchema = (appendState = {}) => {
+    const fields = [];
+    const subnetSelect = [];
+    setState((state) => ({
+      ...state,
+      ...appendState,
+      fields,
+      subnetSelect,
+    }));
+  };
+
+  useEffect(() => {
+    if (routerId && isLoading) {
+      miqSparkleOn();
+      API.get(`/api/network_routers/${routerId}?attributes=cloud_network.cloud_subnets.ids,external_gateway_info`)
+        .then((initialValues) => {
+          initialValues.enable = !!initialValues.extra_attributes.external_gateway_info;
+          if (initialValues.enable) {
+            initialValues.source_nat = initialValues.extra_attributes.external_gateway_info.enable_snat;
+            if (initialValues.network_id) {
+              initialValues.subnet_id = initialValues.extra_attributes.external_gateway_info.external_fixed_ips[0]
+                ? initialValues.extra_attributes.external_gateway_info.external_fixed_ips[0].subnet_id : -1;
+            }
+          }
+          if (initialValues.cloud_network.cloud_subnets.ids) {
+            initialValues.cloud_subnet_id = `${initialValues.cloud_network.cloud_subnets.ids[0]}`;
+          }
+          API.get(
+            `/api/cloud_subnets?expand=resources&attributes=name,ems_ref&filter[]=cloud_network_id=${initialValues.cloud_network_id}`
+          ).then((data) => {
+            let subnets = data.resources;
+            subnets = subnets.map(({ id, name }) => ({ label: name, value: id }));
+            setState((state) => ({
+              ...state,
+              subnets,
+            }));
+          });
+          setState({
+            initialValues,
+            isLoading: false,
+          });
+          API.options(`/api/network_routers/${routerId}`).then(
+            loadSchema({})
+          );
+          miqSparkleOff();
+        });
+    }
+  }, [routerId]);
+
+  const onSubmit = (values) => {
+    const resources = {};
```

Comment:
@GilbertCherrie Is it possible to move some of the logic to a small functions and then move to helpers file?